### PR TITLE
Bulk Load CDK: Unwrap multipart streaming upload

### PR DESCRIPTION
**Pull request by johnny-schmidt**
## What
This is a first step toward separating processing from uploading

* add ObjectStorageClient/S3Client support for uploading a part at a time (leaving the old style in place for now until it can be removed from all destinations/tests)
* add a `BufferingFormattedWriter` that wraps the Avro/Json/Csv/Parquet writers, accumulating the records and yielding complete parts
* change StreamLoader::processRecords so that it feeds the writer and uploads the parts as they're ready

### CODE REVIEW

File: airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt [hunk 1]
```diff
@@ -155,11 +169,73 @@ class ParquetFormattingWriter(
...
+}
+
+class BufferedFormattingWriter<T : OutputStream>(
+    private val writer: ObjectStorageFormattingWriter,
+    private val partSize: Long,
+    private val buffer: ByteArrayOutputStream,
+    private val streamProcessor: StreamProcessor<T>,
+    private val wrappingBuffer: T
+) : ObjectStorageFormattingWriter {
+    override val numCapturedChanges: AtomicLong = writer.numCapturedChanges
```
Comment (tryangul) [inline airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt:214]: What is this for? I only see it used for logging?

File: airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt [hunk 2]
```diff
@@ -155,11 +169,73 @@ class ParquetFormattingWriter(
...
+
+    override fun accept(record: DestinationRecord) {
+        writer.accept(record)
+    }
+
+    fun isDataSufficient(): Boolean {
+        return buffer.size() >= partSize
+    }
```
Comment (tryangul) [inline airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt:222]: Nit: this seems orthogonal to the buffered writer. Since it's called strictly externally, we can easily pull this out.

File: airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt (hunk 1 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt:214]: Yes, it's some of what I added for testing, but if feels like useful information.

File: airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt (hunk 2 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt:222]: Yeah, it should just be a size check.

File: airbyte-cdk/bulk/toolkits/load-s3/src/main/kotlin/io/airbyte/cdk/load/file/s3/S3Client.kt [hunk 3]
```diff
@@ -175,6 +176,33 @@ class S3Client(
         upload.runUsing(block)
         return S3Object(key, bucketConfig)
     }
+
+    override suspend fun startStreamingUpload(
+        key: String,
+        metadata: Map<String, String>
+    ): StreamingUpload<S3Object> {
+        // TODO: Remove permit handling once we control concurrency with # of accumulators
+        if (uploadPermits != null) {
```
Comment (tryangul) [inline airbyte-cdk/bulk/toolkits/load-s3/src/main/kotlin/io/airbyte/cdk/load/file/s3/S3Client.kt:185]: Should we go ahead and move the semaphore upstream to get it out of the client code at least?

**Review by tryangul — APPROVED** [review]
Approved, pending nit change not withstanding

**Comment by johnny-schmidt** [comment]
I also ripped out the logging.
